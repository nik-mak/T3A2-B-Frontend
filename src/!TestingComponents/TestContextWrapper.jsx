import React from 'react'
import AlertsContext from "../contexts/alert";
import CartContext from "../contexts/cart";
import ModalsContext from "../contexts/modals";
import SelectedListingContext from "../contexts/selectedListing";
import CartTotalContext from "../contexts/total";
import UserContext from "../contexts/user";

function TestContextWrapper( {children, user} ) {

  return (
    <SelectedListingContext.Provider value={[]}>
          <ModalsContext.Provider value={[]}>
            <AlertsContext.Provider value={[]}>
              <CartContext.Provider value={[]}>
                <CartTotalContext.Provider value={[]}>
                  <UserContext.Provider
                    value={[user]}
                  >
                    {children}
                  </UserContext.Provider>
                </CartTotalContext.Provider>
              </CartContext.Provider>
            </AlertsContext.Provider>
          </ModalsContext.Provider>
        </SelectedListingContext.Provider>
  )
}

export default TestContextWrapper