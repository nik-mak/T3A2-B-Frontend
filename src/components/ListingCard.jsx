import React, { useContext } from "react";
import UserContext from "../contexts/user";
import api from "../helpers/api";
import useAlerts from "../hooks/useAlerts";

/**
 * ListingCard is a component that handles rendering a listing card to be displayed in the catalogue
 *
 * @param {String} heading the heading to be displayed as the cards title
 * @param {String} size the size to be displayed on the card
 * @param {Number} price the price of the item to be displayed on the card
 * @param {String} imageURL the url link to an image of the item to be displayed on the card
 * @param {itemID} itemID the itemID used to identify the catalogue listing
 *
 * @return {HTML} a listing card
 */
function ListingCard({ heading, size, price, imageURL, itemID }) {
  const user = useContext(UserContext)[0];
  const { addAlert } = useAlerts();

  const { cart } = React.useContext(UserContext);
  const [setCartItems] = cart;

  /**
   * A function that handles adding a listing item to a users cart using the itemID
   */
  function addToCartHandler() {
    api
      .put(`/cart/${itemID}`, itemID)
      .then((response) => {
        addAlert({
          severity: "success",
          message: response.data,
        })
          .then(api.get("/cart"))
          .then(({ data }) => {
            setCartItems(data);
          });
      })
      .catch((response) => {
        if (user === undefined) {
          addAlert({
            severity: "warning",
            message: "Please login to add this item to your cart",
          });
        } else {
          addAlert({ severity: "warning", message: response.message });
        }
      });
  }

  return (
    <div className="m-4 flex flex-col space-y-4 bg-white p-2 font-oswald text-lg shadow-custom">
      <h2>{heading}</h2>
      {/* <BlankImageIcon /> */}
      <img
        src={imageURL}
        alt={heading}
        className="aspect-square w-full object-cover"
      />
      <div className="flex justify-evenly">
        <h3>Size: {size}</h3>
        <h3>
          Price: <span className="font-roman">$</span>
          {price}
        </h3>
      </div>
      <button
        className="w-1/2 self-center rounded-full border border-slate-300 bg-ghost-white py-2 hover:bg-slate-100"
        onClick={addToCartHandler}
      >
        + Add to cart
      </button>
    </div>
  );
}

export default ListingCard;
