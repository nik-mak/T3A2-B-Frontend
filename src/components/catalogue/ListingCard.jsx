import React, { useContext, useState } from "react";
import CartContext from "../../contexts/cart";
import CartTotalContext from "../../contexts/total";
import ModalsContext from "../../contexts/modals";
import SelectedListingContext from "../.././contexts/selectedListing";
import UserContext from "../../contexts/user";
import api from "../../helpers/api";
import useAlerts from "../../hooks/useAlerts";
import HasRole from ".././Auth/HasRole";
import NotLoggedIn from ".././Auth/NotLoggedIn";
import BinIcon from ".././Icons/BinIcon";

/**
 * ListingCard is a component that handles rendering a listing card to be displayed in the catalogue
 *
 * @param {String} heading the heading to be displayed as the cards title
 * @param {String} size the size to be displayed on the card
 * @param {Number} price the price of the item to be displayed on the card
 * @param {String} imageURL the url link to an image of the item to be displayed on the card
 * @param {itemID} itemID the itemID used to identify the catalogue listing
 *
 * @return {HTML} a listing card
 */
function ListingCard({ heading, size, price, imageURL, itemID }) {
  // Defines the user so that roles can be checked when rendering components.
  const { user } = useContext(UserContext);
  // Defines teh cart items to be displayed in the cart
  const [cartItems, setCartItems] = useContext(CartContext)
  const [cartTotal, setCartTotal] = useContext(CartTotalContext)

  // Defines the addAlert method used to create new alerts
  const { addAlert } = useAlerts();

  // Defines the isRemoved property that is used to hide/remove the specific listing from the catalogue if it has been removed
  const [isRemoved, setIsRemoved] = useState(false);

  // defines modalStates and setModalStates using context provider
  const setModalStates = useContext(ModalsContext)[1];
  const setSelectedListing = useContext(SelectedListingContext)[1];

  /**
   * A function that handles removing a listing item from the catalogue using the itemID
   */
  function removeHandler() {
    api
      .delete(`/items/${itemID}`)
      .then(() => {
        addAlert({
          severity: "success",
          message: `Successfully removed item '${heading}'`,
        });
        setIsRemoved(true);
      })
      .catch((error) => {
        addAlert({
          severity: "warning",
          message: `Failed to remove item '${heading}' because of: ` + (error.response.data || error.message),
        });
      });
  }
  /**
   * A function that handles adding a listing item to a users cart using the itemID
   */
  function addToCartHandler() {
    api
      .put(`/cart/${itemID}`, itemID)
      .then((res) => {
        setCartItems([...cartItems, res.data]);
        setCartTotal(cartTotal + res.data.price);
        addAlert({
          severity: "success",
          message: `Successfully added '${heading}' to the your cart!`,
        });
      })
      .catch((response) => {
        if (user === undefined) {
          addAlert({
            severity: "warning",
            message: response.message,
          });
          addAlert({
            severity: "warning",
            message: "Please login to add this item to your cart",
          });
        } 
      });
  }

  return isRemoved ? (
    <></>
  ) : (
    <div className="m-4 flex flex-col space-y-4 bg-white p-2 font-oswald text-lg shadow-custom2">
      <div className="flex flex-row items-center justify-between">
        <h2 className="mr-2">{heading}</h2>
        {/* Displays remove listing if user is staff or admin */}
        <HasRole roles={["admin", "staff"]}>
          <button
            className="mr-2 flex w-[20px] flex-col items-center self-center hover:text-ue-red"
            onClick={removeHandler}
          >
            <BinIcon></BinIcon>
            <p className="font-oswald text-base">remove</p>
          </button>
        </HasRole>
      </div>
      {/* <BlankImageIcon /> */}
      <img
        src={imageURL}
        alt={heading}
        className="aspect-square w-full object-cover"
      />
      <div className="flex justify-evenly">
        <h3>Size: {size}</h3>
        <h3>
          Price: <span className="font-roman">$</span>
          {price}
        </h3>
      </div>
      {/* Displays add to cart button if user is customer */}
      <HasRole roles={["customer"]}>
        <button
          className="w-1/2 self-center rounded-full border border-slate-300 bg-ghost-white py-2 hover:bg-slate-100"
          onClick={addToCartHandler}
        >
          + Add to cart
        </button>
      </HasRole>
      {/* Displays add to cart button if user is not logged in */}
      <NotLoggedIn>
        <button
          className="w-1/2 self-center rounded-full border border-slate-300 bg-ghost-white py-2 hover:bg-slate-100"
          onClick={addToCartHandler}
        >
          + Add to cart
        </button>
      </NotLoggedIn>
      {/* displays manage listing button if user is staff or admin */}
      <HasRole roles={["staff", "admin"]}>
        <button
          className="w-2/3 self-center rounded-full border border-slate-300 bg-ghost-white py-2 hover:bg-slate-100"
          onClick={() => {
            setSelectedListing({
              name: heading,
              size,
              price,
              itemID,
            });
            setModalStates({ modalName: "manageListing", action: "open" });
          }}
        >
          Manage Listing
        </button>
      </HasRole>
    </div>
  );
}

export default ListingCard;
