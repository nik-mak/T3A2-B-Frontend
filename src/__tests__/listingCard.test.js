import renderer from "react-test-renderer";
import ListingCard from "../components/catalogue/ListingCard";
import TestContextWrapper from "../!TestingComponents/TestContextWrapper";


it("Can be rendered", () => {
  const component = renderer.create(
    <TestContextWrapper user={{name:"Test", email:"test@test", role:"staff"}}>
      <ListingCard
        heading="Test Heading"
        size="Test Size"
        price={50}
        imageURL=""
        itemID="1"
      />
    </TestContextWrapper>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
