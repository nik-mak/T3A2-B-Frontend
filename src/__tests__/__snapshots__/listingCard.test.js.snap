// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can be rendered 1`] = `
<div
  className="m-4 flex flex-col space-y-4 bg-white p-2 font-oswald text-lg shadow-custom2"
>
  <div
    className="flex flex-row items-center justify-between"
  >
    <h2
      className="mr-2"
    >
      Test Heading
    </h2>
    <button
      className="mr-2 flex w-[20px] flex-col items-center self-center hover:text-ue-red"
      onClick={[Function]}
    >
      <svg
        fill="none"
        viewBox="0 0 37 44"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M36 7H3.91667L6.83333 43H30.1667L33.0833 7H1M18.5 16V34M25.7917 16L24.3333 34M11.2083 16L12.6667 34M12.6667 7L14.125 1H22.875L24.3333 7"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        />
      </svg>
      <p
        className="font-oswald text-base"
      >
        remove
      </p>
    </button>
  </div>
  <img
    alt="Test Heading"
    className="aspect-square w-full object-cover"
    src=""
  />
  <div
    className="flex justify-evenly"
  >
    <h3>
      Size: 
      Test Size
    </h3>
    <h3>
      Price: 
      <span
        className="font-roman"
      >
        $
      </span>
      50
    </h3>
  </div>
  <button
    className="w-2/3 self-center rounded-full border border-slate-300 bg-ghost-white py-2 hover:bg-slate-100"
    onClick={[Function]}
  >
    Manage Listing
  </button>
</div>
`;
